content.go

```diff
@@ -76,25 +90,106 @@ func Resolve(ctx context.Context, target ReadOnlyTarget, ref string, opts Resolv
-			return selectPlatform(ctx, proxy, desc, opts.TargetPlatform)
+			return platform.SelectManifest(ctx, proxy, desc, opts.TargetPlatform)
 		default:
 			return ocispec.Descriptor{}, fmt.Errorf("%s: %s: %w", desc.Digest, desc.MediaType, errdef.ErrUnsupported)
 		}
 	}
 
-	desc, err := target.Resolve(ctx, ref)
+	desc, err := target.Resolve(ctx, reference)
 	if err != nil {
 		return ocispec.Descriptor{}, err
 	}
+	return platform.SelectManifest(ctx, target, desc, opts.TargetPlatform)
+}
+
+// DefaultFetchOptions provides the default FetchOptions.
+var DefaultFetchOptions = FetchOptions{}
+
+// FetchOptions contains parameters for oras.Fetch.
+type FetchOptions struct {
+	// ResolveOptions contains parameters for resolving reference.
+	ResolveOptions
+}
+
+// Fetch fetches the content identified by the reference.
+func Fetch(ctx context.Context, target ReadOnlyTarget, reference string, opts FetchOptions) (ocispec.Descriptor, io.ReadCloser, error) {
+	if opts.TargetPlatform == nil {
+		if refFetcher, ok := target.(registry.ReferenceFetcher); ok {
+			return refFetcher.FetchReference(ctx, reference)
+		}
+
+		desc, err := target.Resolve(ctx, reference)
+		if err != nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+		rc, err := target.Fetch(ctx, desc)
+		if err != nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+		return desc, rc, nil
+	}
+
+	proxy := cas.NewProxy(target, cas.NewMemory())
+	desc, err := resolve(ctx, target, proxy, reference, opts.ResolveOptions)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	// if the content exists in cache, fetch it from cache
+	// otherwise fetch without caching
+	proxy.StopCaching = true
+	rc, err := proxy.Fetch(ctx, desc)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	return desc, rc, nil
+}
+
+// DefaultFetchBytesOptions provides the default FetchBytesOptions.
+var DefaultFetchBytesOptions = FetchBytesOptions{
+	SizeLimit: int64(1 << 22), // 4 MiB
+}
+
+// FetchBytesOptions contains parameters for oras.FetchBytes.
+type FetchBytesOptions struct {
+	FetchOptions
+	// SizeLimit limits the max size of the fetched content.
+	// If SizeLimit is not specified, or the specified value is less than or
+	// equal to 0, it will be considered as infinity.
+	SizeLimit int64
+}
+
+// FetchBytes fetches the content bytes identified by the reference.
+func FetchBytes(ctx context.Context, target ReadOnlyTarget, reference string, opts FetchBytesOptions) (ocispec.Descriptor, []byte, error) {
+	desc, rc, err := Fetch(ctx, target, reference, opts.FetchOptions)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	defer rc.Close()
+
+	var bytes []byte
+	if opts.SizeLimit > 0 {
+		if desc.Size > opts.SizeLimit {
+			return ocispec.Descriptor{}, nil, fmt.Errorf(
+				"content size %v exceeds max size limit %v: %w",
+				desc.Size,
+				opts.SizeLimit,
+				errdef.ErrSizeExceedsLimit)
+		}
+		bytes, err = content.ReadAll(io.LimitReader(rc, opts.SizeLimit), desc)
```

Comment:
`content.ReadAll` takes care of the size limit.
```suggestion
		bytes, err = content.ReadAll(rc, desc)
```